Parallel drain all nodegroups

hintofbasil opened the pull request:
Currently all nodegroups are drained in serial but the nodes in each node group
are drained in parallel.  This change updates this logic to drain all nodes
across all node groups in parallel - up a maximum defined by the --parallel
flag.

This significantly speeds up draining for clusters with a large number of
small node groups.  Especially if some of the nodes in these node groups
take longer to drain than others.

This has the side affect of cordoning all nodes in all node groups before
beginning the actual draining process.  This should be the preferred behaviour
but may come as a surprise to some users - especially those that assume the
cordoning is actually the node being drained.


I attempted to add new tests for `pkg/actions/nodegroup/drain.go` however this
ended up with a rather unwieldy refactor and I decided against it for this change.
If you have any advice for a reasonable way to implement these tests then I'm
happy to add them.


Logs from testing
```
2022-04-21 23:28:52 [ℹ]  eksctl version 0.95.0-dev+61bd17de.2022-04-21T23:26:39Z
2022-04-21 23:28:52 [ℹ]  using region eu-west-1
2022-04-21 23:28:53 [ℹ]  comparing 2 nodegroups defined in the given config ("/tmp/cellsdev-1-eu-west-1a-1.yaml") against remote state
2022-04-21 23:28:55 [ℹ]  6 nodegroup(s) present in the config file (mixed-instances-12xl-53a517ef,mixed-instances-4xl-523169f7,mixed-instances-8xl-02c168d2,nodes-default-gateway-ingress-e0d2c388,nodes-ingress-adb3f800,nodes-public-ingress-51379634) will be excluded
2022-04-21 23:28:55 [ℹ]  3 nodegroups (mixed-instances-12xl-53a517ef-test, mixed-instances-4xl-523169f7-test, mixed-instances-8xl-02c168d2-test) were included (based on the include/exclude rules)
2022-04-21 23:28:55 [ℹ]  will drain 3 nodegroup(s) in cluster "cellsdev-1-eu-west-1a-1"
2022-04-21 23:28:55 [ℹ]  starting parallel draining, max in-flight of 3
2022-04-21 23:28:56 [ℹ]  cordon node "ip-172-16-20-6.eu-west-1.compute.internal"
2022-04-21 23:28:56 [ℹ]  cordon node "ip-172-16-10-183.eu-west-1.compute.internal"
2022-04-21 23:28:56 [ℹ]  cordon node "ip-172-16-30-16.eu-west-1.compute.internal"
2022-04-21 23:29:40 [✔]  drained all nodes: [ip-172-16-20-6.eu-west-1.compute.internal]
2022-04-21 23:29:40 [✔]  drained all nodes: [ip-172-16-10-183.eu-west-1.compute.internal]
2022-04-21 23:30:05 [✔]  drained all nodes: [ip-172-16-30-16.eu-west-1.compute.internal]
2022-04-21 23:30:05 [ℹ]  will delete 3 nodegroups from cluster "cellsdev-1-eu-west-1a-1"
2022-04-21 23:30:07 [ℹ]  1 task: { 
3 parallel tasks: { delete nodegroup "mixed-instances-12xl-53a517ef-test" [async], delete nodegroup "mixed-instances-4xl-523169f7-test" [async], delete nodegroup "mixed-instances-8xl-02c168d2-test" [async] 
} }2022-04-21 23:30:07 [ℹ]  will delete stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-523169f7-test"
2022-04-21 23:30:07 [ℹ]  will delete stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-02c168d2-test"
2022-04-21 23:30:07 [ℹ]  will delete stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-12xl-53a517ef-test"
2022-04-21 23:30:07 [✔]  deleted 3 nodegroup(s) from cluster "cellsdev-1-eu-west-1a-1"
```

### Description

<!--
Please explain the changes you made here.

Help your reviewers my guiding them through your key changes,
implementation decisions etc.
You can even include snippets of output or screenshots.

A good, clear description == a faster review :)
-->

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

Skarlso commented:
Hey @hintofbasil :)

This now removed the draining of nodes in parallel and changed it to draining nodegroups in parallel. What about a nodegroup that has 50-60 nodes? They would be drained sequentially.

Also, this changes the current default behaviour, so I would instead think that this should be an opt-in feature? WDYT?

Also, also, this should come with a warning because if people are trying to A-B deployment and using eksctl to delete a deployment. They get all of the nodegroups drained simultaneously, which might disrupt service availability.

Also, also, also (:D) changes how a setting previously used for parallel node draining is used, which breaks backwards compatibility. This flag might be used already for X but will now do Y if that makes sense? :)

hintofbasil commented:
Hi @Skarlso


> This now removed the draining of nodes in parallel and changed it to draining nodegroups in parallel. What about a nodegroup that has 50-60 nodes? They would be drained sequentially.

I maybe should have been clearer in the PR description.  This change is still limited by  the `--parallel` flag so even across node groups it will drain a maximum of `--parallel` nodes at one time.   This defaults to `1` so for users that haven't opted in to parallel draining this won't change their draining behaviour.

> Also, this changes the current default behaviour, so I would instead think that this should be an opt-in feature? WDYT?

For people with multiple node groups this will now cordon all the nodes at the start.  This is the current behaviour for single node group clusters - and I can't see a use case where that isn't the desired behaviour for multiple node group clusters.

Currently as one node group drains the evicted pods get scheduled onto nodes in node groups which are about to be drained.  This means we end up shuffling pods around multiple times significantly extending the drain time.  Cordoning all the nodes at the start simply means the evicted pods get scheduled onto nodes which aren't due to be drained much earlier.

But I am happy to put that functionality behind a flag if you think it's better.

> Also, also, this should come with a warning because if people are trying to A-B deployment and using eksctl to delete a deployment. They get all of the nodegroups drained simultaneously, which might disrupt service availability.

Do you still think this is needed given the default parallel level of `1`.  Anyone opting into a higher parallel rate should expect this (or have PDBs deployed).

Skarlso commented:
>I maybe should have been clearer in the PR description. This change is still limited by the --parallel flag so even across node groups it will drain a maximum of --parallel nodes at one time. This defaults to 1 so for users that haven't opted in to parallel draining this won't change their draining behaviour.

No, you're right. I read that incorrectly. I see that the feature is still there. :)

>For people with multiple node groups this will now cordon all the nodes at the start. This is the current behaviour for single node group clusters - and I can't see a use case where that isn't the desired behaviour for multiple node group clusters.

Hmm, true.

>Do you still think this is needed given the default parallel level of 1. Anyone opting into a higher parallel rate should expect this (or have PDBs deployed).

Maybe it's fine. :D

Skarlso reviewed (commented):
This looks okay to me.

hintofbasil commented:
> Please bear with us, the team has limited capacity. 🤗

Of course, there is no rush here.

hintofbasil commented:
Are you fine with a rebase and force push to fix the conflicts? Or would you prefer a pull?

Skarlso commented:
Force pushing is fine.

hintofbasil commented:
Rebased.

Let me know if another manual test is required.

Himangini reviewed (approved):
LGTM, if you can post the output of manual tests with the latest version that'll be awesome.

nikimanoledaki reviewed (approved):
Just one thing from my side 😄

hintofbasil commented on pkg/actions/nodegroup/drain.go:
```diff
@@ -20,13 +24,35 @@ type DrainInput struct {
+
+		g, ctx := errgroup.WithContext(ctx)
+		for _, nodegroup := range input.NodeGroups {
+			nodegroup := nodegroup
```
With this line the linter adds a warning

```
loop variable nodegroup captured by func literal loopclosure
```

hintofbasil commented:
Node creation test output

```
❯ ./eksctl create nodegroup --config-file /tmp/cellsdev-1-eu-west-1a-1.yaml
2022-05-26 13:21:30 [ℹ]  nodegroup "nodes-ingress-2f8bd6f9" will use "ami-0440c10a3f77514d8" [AmazonLinux2/1.21]
2022-05-26 13:21:30 [ℹ]  using SSH public key "/tmp/cells.pub" as "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-nodes-ingress-2f8bd6f9-91:90:3e:d2:6c:c5:4f:ca:80:00:8a:24:a7:5e:b8:e0"
2022-05-26 13:21:31 [ℹ]  nodegroup "nodes-public-ingress-b38135fb" will use "ami-0440c10a3f77514d8" [AmazonLinux2/1.21]
2022-05-26 13:21:31 [ℹ]  using SSH public key "/tmp/cells.pub" as "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-nodes-public-ingress-b38135fb-91:90:3e:d2:6c:c5:4f:ca:80:00:8a:24:a7:5e:b8:e0"
2022-05-26 13:21:31 [ℹ]  nodegroup "mixed-instances-12xl-159f5dd8" will use "ami-0440c10a3f77514d8" [AmazonLinux2/1.21]
2022-05-26 13:21:31 [ℹ]  using SSH public key "/tmp/cells.pub" as "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-12xl-159f5dd8-91:90:3e:d2:6c:c5:4f:ca:80:00:8a:24:a7:5e:b8:e0"
2022-05-26 13:21:31 [ℹ]  nodegroup "mixed-instances-4xl-eb27c1d4-test" will use "ami-0440c10a3f77514d8" [AmazonLinux2/1.21]
2022-05-26 13:21:31 [ℹ]  using SSH public key "/tmp/cells.pub" as "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test-91:90:3e:d2:6c:c5:4f:ca:80:00:8a:24:a7:5e:b8:e0"
2022-05-26 13:21:32 [ℹ]  nodegroup "mixed-instances-8xl-aa39b7d8-test" will use "ami-0440c10a3f77514d8" [AmazonLinux2/1.21]
2022-05-26 13:21:32 [ℹ]  using SSH public key "/tmp/cells.pub" as "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test-91:90:3e:d2:6c:c5:4f:ca:80:00:8a:24:a7:5e:b8:e0"
2022-05-26 13:21:32 [ℹ]  nodegroup "nodes-default-gateway-ingress-d6564902" will use "ami-0440c10a3f77514d8" [AmazonLinux2/1.21]
2022-05-26 13:21:32 [ℹ]  using SSH public key "/tmp/cells.pub" as "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-nodes-default-gateway-ingress-d6564902-91:90:3e:d2:6c:c5:4f:ca:80:00:8a:24:a7:5e:b8:e0"
2022-05-26 13:21:35 [ℹ]  6 existing nodegroup(s) (mixed-instances-12xl-159f5dd8,mixed-instances-4xl-eb27c1d4,mixed-instances-8xl-aa39b7d8,nodes-default-gateway-ingress-d6564902,nodes-ingress-2f8bd6f9,nodes-public-ingress-b38135fb) will be excluded
2022-05-26 13:21:35 [ℹ]  2 nodegroups (mixed-instances-4xl-eb27c1d4-test, mixed-instances-8xl-aa39b7d8-test) were included (based on the include/exclude rules)
2022-05-26 13:21:35 [ℹ]  will create a CloudFormation stack for each of 2 managed nodegroups in cluster "cellsdev-1-eu-west-1a-1"
2022-05-26 13:21:35 [ℹ]
2 sequential tasks: { fix cluster compatibility, 1 task: {
2 parallel tasks: { create managed nodegroup "mixed-instances-4xl-eb27c1d4-test", create managed nodegroup "mixed-instances-8xl-aa39b7d8-test"
} }
}
2022-05-26 13:21:35 [ℹ]  checking cluster stack for missing resources
2022-05-26 13:21:38 [ℹ]  cluster stack has all required resources
2022-05-26 13:21:40 [ℹ]  building managed nodegroup stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:21:40 [ℹ]  deploying stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:21:40 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:21:41 [ℹ]  building managed nodegroup stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:21:41 [ℹ]  deploying stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:21:41 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:22:10 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:22:12 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:23:02 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:23:10 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:24:00 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:24:18 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:24:43 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:25:17 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:25:50 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:26:49 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:27:25 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:27:30 [ℹ]  waiting for CloudFormation stack "eksctl-cellsdev-1-eu-west-1a-1-nodegroup-mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:27:32 [ℹ]
2 parallel tasks: { suspend ASG processes for nodegroup mixed-instances-4xl-eb27c1d4-test, suspend ASG processes for nodegroup mixed-instances-8xl-aa39b7d8-test
}
2022-05-26 13:27:34 [ℹ]  suspended ASG processes [AZRebalance] for mixed-instances-8xl-aa39b7d8-test
2022-05-26 13:27:34 [ℹ]  suspended ASG processes [AZRebalance] for mixed-instances-4xl-eb27c1d4-test
2022-05-26 13:27:34 [✔]  created 0 nodegroup(s) in cluster "cellsdev-1-eu-west-1a-1"
2022-05-26 13:27:34 [ℹ]  nodegroup "mixed-instances-4xl-eb27c1d4-test" has 5 node(s)
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-12-242.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-15-4.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-2-249.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-30-219.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-8-231.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  waiting for at least 5 node(s) to become ready in "mixed-instances-4xl-eb27c1d4-test"
2022-05-26 13:27:34 [ℹ]  nodegroup "mixed-instances-4xl-eb27c1d4-test" has 5 node(s)
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-12-242.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-15-4.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-2-249.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-30-219.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-8-231.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  nodegroup "mixed-instances-8xl-aa39b7d8-test" has 5 node(s)
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-12-231.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-23-103.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-26-128.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-8-174.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  node "ip-172-16-8-4.eu-west-1.compute.internal" is ready
2022-05-26 13:27:34 [ℹ]  waiting for at least 5 node(s) to become ready in "mixed-instances-8xl-aa39b7d8-test"
2022-05-26 13:27:35 [ℹ]  nodegroup "mixed-instances-8xl-aa39b7d8-test" has 5 node(s)
2022-05-26 13:27:35 [ℹ]  node "ip-172-16-12-231.eu-west-1.compute.internal" is ready
2022-05-26 13:27:35 [ℹ]  node "ip-172-16-23-103.eu-west-1.compute.internal" is ready
2022-05-26 13:27:35 [ℹ]  node "ip-172-16-26-128.eu-west-1.compute.internal" is ready
2022-05-26 13:27:35 [ℹ]  node "ip-172-16-8-174.eu-west-1.compute.internal" is ready
2022-05-26 13:27:35 [ℹ]  node "ip-172-16-8-4.eu-west-1.compute.internal" is ready
2022-05-26 13:27:35 [✔]  created 2 managed nodegroup(s) in cluster "cellsdev-1-eu-west-1a-1"
2022-05-26 13:27:37 [ℹ]  checking security group configuration for all nodegroups
2022-05-26 13:27:37 [ℹ]  all nodegroups have up-to-date cloudformation templates
```
